src/utils/src/ExponentialBackoffUtils.c

```diff
@@ -0,0 +1,222 @@
+    UINT64 power = 0, waitTime = 0;
+
+    CHK(pWaitTime != NULL, STATUS_NULL_ARG);
+
+    CHK_STATUS(computePower(DEFAULT_KVS_EXPONENTIAL_FACTOR, pRetryState->currentRetryCount, &power));
+    waitTime = power * pRetryConfig->retryFactorTime;
+
+    *pWaitTime = waitTime;
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS validateAndUpdateExponentialBackoffStatus(PExponentialBackoffState pExponentialBackoffState) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    switch (pExponentialBackoffState->status) {
+        case BACKOFF_NOT_STARTED:
+            pExponentialBackoffState->status = BACKOFF_IN_PROGRESS;
+            DLOGD("Status changed from BACKOFF_NOT_STARTED to BACKOFF_IN_PROGRESS");
+            break;
+        case BACKOFF_IN_PROGRESS:
+            // Do nothing
+            DLOGD("Current status is BACKOFF_IN_PROGRESS");
+            break;
+        case BACKOFF_TERMINATED:
+            DLOGD("Cannot execute exponentialBackoffBlockingWait. Current status is BACKOFF_TERMINATED");
+            CHK(FALSE, STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE);
+            // No 'break' needed since CHK(FALSE, ...) will always jump to CleanUp
+        default:
+            DLOGD("Cannot execute exponentialBackoffBlockingWait. Unexpected state [%"PRIu64"]", pExponentialBackoffState->status);
+            CHK(FALSE, STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE);
+    }
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS exponentialBackoffBlockingWait(PExponentialBackoffState pRetryState) {
+    ENTERS();
+    PExponentialBackoffConfig pRetryConfig = NULL;
+    UINT64 currentTime = 0, waitTime = 0, jitter = 0;
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pRetryState != NULL, STATUS_NULL_ARG);
+
+    CHK_STATUS(validateAndUpdateExponentialBackoffStatus(pRetryState));
+
+    pRetryConfig = &(pRetryState->exponentialBackoffConfig);
+
+    // If retries is exhausted, return error to the application
+    CHK(pRetryConfig->maxRetryCount == KVS_INFINITE_EXPONENTIAL_RETRIES ||
+         pRetryConfig->maxRetryCount > pRetryState->currentRetryCount, STATUS_EXPONENTIAL_BACKOFF_RETRIES_EXHAUSTED);
+
+    // check the last retry time. If the difference between last retry and current time is greater than
+    // minTimeToResetRetryState, then reset the retry state
+    //
+    // Proceed if this is not the first retry
+    currentTime = GETTIME();
+    if (pRetryState->currentRetryCount != 0 &&
+        currentTime > pRetryState->lastRetryWaitTime && // sanity check
+        currentTime - pRetryState->lastRetryWaitTime > pRetryConfig->minTimeToResetRetryState) {
+        CHK_STATUS(resetExponentialBackoffRetryState(pRetryState));
+        pRetryState->status = BACKOFF_IN_PROGRESS;
+    }
+
+    // Bound the exponential curve to maxWaitTime. Once we reach
+    // maxWaitTime, then we always wait for maxWaitTime time
+    // till the state is reset.
+    if (pRetryState->lastRetryWaitTime == pRetryConfig->maxWaitTime) {
```

Comment:
Is it possible for `pRetryState->lastRetryWaitTime > pRetryConfig->maxWaitTime`
?  Seems like it would be an error, not intended, do we need to check for that is there a way that could happen?